binary_search_tree/tree.py

```diff
@@ -9,51 +9,138 @@ def __init__(self, key, val = None):
+        
+        if current_node == None:
+            return None
+
+        if current_node.key == key:
+            return current_node.value
+
+        if current_node.key > key:
+            return self.find_helper(current_node.left, key)
+        else:
+            return self.find_helper(current_node.right, key)
+
+
+    # Time Complexity: O(log(n)) 
+    # Space Complexity: O(1)
     def find(self, key):
-        pass
+        return self.find_helper(self.root, key)
+
 
-    # Time Complexity: 
-    # Space Complexity: 
+    def inorder_helper(self, current_node, inorder_return):
+        if current_node == None:
+            return inorder_return
+
+        self.inorder_helper(current_node.left, inorder_return)
+        inorder_return.append({ 
+            "key": current_node.key,
+            "value": current_node.value
+        })
+        self.inorder_helper(current_node.right, inorder_return)
+
+        return inorder_return 
+        
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def inorder(self):
-        pass
+        inorder_return = []
+        return self.inorder_helper(self.root, inorder_return)
 
-    # Time Complexity: 
-    # Space Complexity:     
+
+    def preorder_helper(self, current_node, preorder_return):
+        if current_node == None:
+            return preorder_return
+
+        preorder_return.append({ 
+            "key": current_node.key,
+            "value": current_node.value
+        })
+        self.preorder_helper(current_node.left, preorder_return)        
+        self.preorder_helper(current_node.right, preorder_return)
+
+        return preorder_return
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)  
     def preorder(self):
-        pass
+        preorder_return = []
+        return self.preorder_helper(self.root, preorder_return)
+
+
+    def postorder_helper(self, current_node, postorder_return):
+        if current_node == None:
+            return postorder_return
+
+        self.postorder_helper(current_node.left, postorder_return)        
+        self.postorder_helper(current_node.right, postorder_return)
+        postorder_return.append({ 
+            "key": current_node.key,
+            "value": current_node.value
+        })
+
+        return postorder_return
 
-    # Time Complexity: 
-    # Space Complexity:     
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)  
     def postorder(self):
-        pass
+        postorder_return = []
+        return self.postorder_helper(self.root, postorder_return)
+
+
+    def height_helper(self, current_node):
+        if current_node == None:
+            return 0
+        
+        left_node = self.height_helper(current_node.left)
+        right_node = self.height_helper(current_node.right)
 
-    # Time Complexity: 
-    # Space Complexity:     
+        return max(left_node, right_node) + 1
+
+    # Time Complexity: O(n) 
+    # Space Complexity: O(1)     
     def height(self):
```

Comment:
👍  The space complexity is O(h) where h is the height of the tree because of the recursive call stack.